### PR TITLE
Shared pool among clients and fixed pool resources release

### LINKED ISSUE #16: how was the connection pool closed?
https://github.com/RediSearch/redisearch-go/blob/master/redisearch/client.go#L42
i suppose this resource requires disposal on system reload or sth? to avoid idle conn pileup at redis side.

### PR DESCRIPTION
**Pull request by filipecosta90**
- [add] added option to create clients with shared pool, via ` NewClientFromPool(pool *redis.Pool, name string) *Client`
- [fix] added option to releases the resources used by the pool ( single, and multi-host ) via `Close()`

fixes #16, #9 

simple example of two clients sharing the pool:
```
        pool := &redis.Pool{Dial: func() (redis.Conn, error) {
		return redis.Dial("tcp", host, redis.DialPassword(password))
	}, MaxIdle: maxConns}
	client1 := NewClientFromPool(pool,"index1")
	client2 := NewClientFromPool(pool,"index2")
````

### CODE REVIEW

**Comment by filipecosta90** [comment]
@ericfengchao can you review this one as well? Would like your input but was not able to mark you as reviewer

File: redisearch/pool.go [hunk 1]
```diff
@@ -65,3 +65,18 @@ func (p *MultiHostPool) Get() redis.Conn {
...
+
+func (p *MultiHostPool) Close() (err error) {
+	p.Lock()
+	defer p.Unlock()
+	for _, host := range p.hosts {
+		pool, found := p.pools[host]
+		if found {
+			err = pool.Close()
+			if err != nil {
+				return
```
Comment (ericfengchao) [inline redisearch/pool.go:77]: if any pool close() errors out, you want to return and not proceed with the rest pools? and why not just ranging over `p.pools` instead? seems more straight forward to me

File: redisearch/pool.go [hunk 2]
```diff
@@ -65,3 +65,18 @@ func (p *MultiHostPool) Get() redis.Conn {
 	return pool.Get()
 
 }
+
+func (p *MultiHostPool) Close() (err error) {
```
Comment (ericfengchao) [inline redisearch/pool.go:69]: not a big fan of the named returned variable thing, unless we want to defer-change the value of it. no biggie, just personal preferences.

File: redisearch/pool_test.go [hunk 3]
```diff
@@ -21,13 +16,17 @@ func TestNewMultiHostPool(t *testing.T) {
...
+				got := NewMultiHostPool(tt.args.hosts)
+				conn := got.Get()
+				if conn == nil {
+					t.Errorf("NewMultiHostPool() = got nil connection")
+				}
+				err := got.Close()
+				assert.Nil(t, err)
+			})
+		}
 	}
```
Comment (ericfengchao) [inline redisearch/pool_test.go:31]: could you also add one more test for the newly added `Close() error` interface?

File: redisearch/pool.go (hunk 2 above)
Comment (filipecosta90) [reply redisearch/pool.go:69]: I believe it simplifies the method. specifically after the last change. please revise last changes

**Comment by filipecosta90** [comment]
@itamarhaber and @ericfengchao can you double-check the changes :)

**Comment by filipecosta90** [comment]
> Although the change to Close will only report the last error I'm good with that (is returning an array of errors something that's done in golang anyway?)

Now that you've mentioned it @itamarhaber I believe the best way is to keep the function signature but to generate an error with all details of error occurrences (I'll change it and return back in a comment to see if you guys like that idea ).

File: redisearch/pool.go [hunk 4]
```diff
@@ -65,3 +66,20 @@ func (p *MultiHostPool) Get() redis.Conn {
...
+	p.Lock()
+	defer p.Unlock()
+	for host, pool := range p.pools {
+		poolErr := pool.Close()
+		//preserve pool error if not nil but continue
```
Comment (filipecosta90) [inline redisearch/pool.go:75]: @itamarhaber what do you think about the following solution?

Comment (itamarhaber) [reply redisearch/pool.go:75]: Lovely 🔨

**Comment by filipecosta90** [comment]
> lgtm 100 but ci was failing tho

It's failing only due to coverage difference -0.06% with the added error concatenation. I believe we can merge it :+1:

**Pull request merged**